### PR TITLE
FEN string parsing support

### LINKED ISSUE #2: Parsing FEN
It would be good to be able to parse [FEN notation](https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation) for deriving board states. We might also want to include a standard starting position for the board somewhere in the code, as it'll be the most common position in the game.

### PR DESCRIPTION
**Pull request by pyroMechanical**
closes #2

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -11,4 +13,32 @@ mod tests {
         let result = add(2, 2);
         assert_eq!(result, 4);
     }
+
+    #[test]
+    fn fen_starting_board() {
```
Comment (rubyr) [inline src/lib.rs]: I would probably add the FEN tests to the FEN module, so they're easier to find.

File: src/parse.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
+use nom::character::complete::{one_of, char, multispace1, u8, u64};
+use nom::bytes::complete::tag;
+use nom::branch::alt;
+use nom::combinator::{recognize, opt};
+use nom::sequence::{tuple, terminated, preceded, pair};
+use nom::IResult;
+use nom::multi::{count, many1};
+
+#[derive(PartialEq, Debug)]
+#[allow(dead_code)]
```
Comment (rubyr) [inline src/parse.rs]: This is okay for now, but good to note that we shouldn't be allowing things like this in the future.

File: src/parse.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    type Error = u8;
+    fn try_from(rank: u8) -> Result<File, Self::Error>{
+        match rank {
+            0..=7 => Ok(unsafe{std::mem::transmute(rank)}),
+            x => Err(x)
+        }
+    }
+}
+
+pub fn fen_to_game(input: &str) -> Result<ParsedGameState, FenError>{
```
Comment (rubyr) [inline src/parse.rs]: We might want to break this out into several smaller functions, certainly at least one for piece positioning and perhaps another one for en passant.

File: src/parse.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    one_of("wb")(input)
+}
+
+fn fen_castling(input: &str) -> IResult<&str, &str> {
+    alt((tag("-"),recognize(many1(one_of("KQkq")))))(input)
+}
+
+fn fen_en_passant_target(input: &str) -> IResult<&str, &str> {
+    alt((tag("-"), recognize(pair(one_of("abcdefgh"), one_of("12345678")))))(input)
+}
```
Comment (rubyr) [inline src/parse.rs]: A minor nitpick, but I think defining these after they're used can make things a little tricky. Feel free to ignore though

File: src/parse.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    match tuple((fen_board, preceded(multispace1, fen_active_player), preceded(multispace1, fen_castling), preceded(multispace1, fen_en_passant_target), preceded(multispace1, u8), preceded(multispace1, u64)))(input) {
+        Err(e) => return Err(FenError::ParseErr(e)),
+        Ok((_, (parsed_board, parsed_active_player, parsed_castling_rights, parsed_en_passant_target, half_turn_clock, full_turn_clock))) => {
+            let mut piece_position: [[char;8];8] = [['.';8];8];
+            for (num, row) in parsed_board.into_iter().enumerate(){
+                let mut result_row: Vec<char> = vec![];
+                for c in row {
+                    match c {
+                        '1'..='8' => {
+                            let x = (c as u8) - b'0';
```
Comment (rubyr) [inline src/parse.rs]: this line's a little confusing at first glance, why not just parse the number out of the char?

**Review by rubyr — CHANGES_REQUESTED** [review]
This looks good, but there's a few things to address with it. My main concern is the tests, those should be in the parse module rather than lib.rs.

File: src/lib.rs (hunk 1 above)
Comment (pyroMechanical) [reply src/lib.rs]: good idea!

File: src/parse.rs (hunk 5 above)
Comment (pyroMechanical) [reply src/parse.rs]: :woman_facepalming: I completely forgot that you could do that. will change

File: src/parse.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                black_queenside: true,
+                white_kingside: true,
+                white_queenside: true
+            },
+            en_passant_target: Some((Rank::E, File::Three)),
+            half_turn_clock: 0,
+            full_turn_clock: 1
+        })
+    )
+}
```
Comment (rubyr) [inline src/parse.rs]: This is good, though we should include it in a `test` mod so it doesn't get included in builds

File: src/parse.rs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    let mut piece_position = [['.'; 8]; 8];
+    for (num, row) in parsed_board.into_iter().enumerate() {
+        let mut result_row: Vec<char> = vec![];
+        for c in row {
+            match c {
+                '1'..='9' => {
+                    let x: u8 = c.to_digit(10).unwrap() as u8;
+                    result_row.append(&mut vec!['.'; x.into()]);
+                }
+                'r' | 'n' | 'b' | 'q' | 'k' | 'p' | 'R' | 'N' | 'B' | 'Q' | 'K' | 'P' => {
```
Comment (rubyr) [inline src/parse.rs]: minor nitpick but could match on c.to_lower() rather than needing duplicate chars here

**Review by rubyr — APPROVED** [review]
This looks much nicer! Minor nitpick with the tests, and a small comment on piece matching, but otherwise looks good to me

**Pull request merged**